src/Dialect/ONNX/Rewrite.cpp

```diff
@@ -1043,6 +1043,227 @@ class ReplaceUnsqueezeOfExpandRewritePattern
+// ```
+// with
+// ```
+// %0 = onnx.Reshape(%X, %new_shape) {allowzero}
+// ```
+// where `%new_shape` is computed from `%shape1` and `%shape2` if possible.
+//
+// We only consider `allowzero=0` in this pattern.
+//
+// # Shape conditions
+//
+// According to ONNX specification for Reshape
+// (https://onnx.ai/onnx/operators/onnx__Reshape.html#):
+// - At most one dimension of the new shape can be -1. In this case, the value
+// is inferred from the size of the tensor and the remaining dimensions
+// - Dimension could also be 0. In this case,
+//   - if allowzero = 0, the actual dimension value is unchanged;
+//   - if allowzero = 1, the dimension will be set explicitly to zero.
+// - If allowzero = 1, it is invalid for the specified shape to contain both a
+// zero value and -1
+//
+// # Combining rules
+//
+// In this pattern, we use the following terms for values in a shape tensor:
+// 0, -1, and L (a literal).
+//
+// These are the rules to combine two values:
+//  (1st)  : (2nd)  => (result)
+//   0     : 0      => 0
+//   0     : L      => L
+//   0     : -1     => -1
+//
+//  -1     : 0      => -1
+//  -1     : L      => L
+//  -1     : -1     => -1
+//
+//   L     : 0      => L
+//   L     : L      => L
+//   L     : -1     => -1
+//
+// To produce a new shape, we combine each value one by one from left to right.
+//
+// Example (allowzero = 0):
+// Ex1. 1st: [0, -1, 0, 5], 2nd: [0, -1, 0] => [0, -1, 0]
+// Ex2. 1st: [0, -1, 0, 5], 2nd: [5, -1, 0] => [5, -1, 0]
+// Ex3. 1st: [0, -1, 0, 5], 2nd: [-1, 0, 0] => [-1, -1, 0]
+// Ex4. 1st: [0, -1, 0, 5], 2nd: [0, 0, 5] => [0, -1, 5]
+// Ex5. 1st: [0, -1, 5, 0], 2nd: [-1, 5, 0] => [-1, 5, 5]
+//
+// After combining two shapes, we check if the result shape is valid or not
+// according to the shape conditions. If it is invalid, the two ReshapeOps are
+// not combined. For example, the output shape in Ex3 is invalid because of two
+// -1s.
+//
+class FuseTwoReshapesPattern : public OpRewritePattern<ONNXReshapeOp> {
+public:
+  using OpRewritePattern<ONNXReshapeOp>::OpRewritePattern;
+
+  FuseTwoReshapesPattern(MLIRContext *context) : OpRewritePattern(context) {}
+
+  LogicalResult matchAndRewrite(
+      ONNXReshapeOp secondReshapeOp, PatternRewriter &rewriter) const override {
+    // Second Reshape.
+    Operation *op = secondReshapeOp.getOperation();
+    Value secondData = secondReshapeOp.getData();
+    Value secondShape = secondReshapeOp.getShape();
+    int64_t secondAllowZero = secondReshapeOp.getAllowzero();
+    if (secondAllowZero != 0)
+      return rewriter.notifyMatchFailure(op, [&](::mlir::Diagnostic &diag) {
+        diag << "Does not support AllowZero != 0";
+      });
+
+    // First Reshape.
+    if (!definedBy<ONNXReshapeOp>(secondData))
+      return rewriter.notifyMatchFailure(op, [&](::mlir::Diagnostic &diag) {
+        diag << "The input data is not defined by a Reshape";
+      });
+    auto firstReshapeOp = cast<ONNXReshapeOp>(secondData.getDefiningOp());
```

Comment:
I recently learned that getDefiningOp can take a template argument, so you can say
```
auto firstReshapeOp = secondData.getDefiningOp<ONNXReshapeOp>();
if (!firstReshapeOp) return ...;
```

Reply:
Yeah, I saw it yesterday too.